### PR TITLE
add unit test for pkg/fs/fs.go

### PR DESCRIPTION
**Pull request by Hellcatlk**
Signed-off-by: Zouyu <zouy.fnst@cn.fujitsu.com>

### CODE REVIEW

File: pkg/fs/fs_test.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+	"os"
+	"testing"
+)
+
+func TestTempDir(t *testing.T) {
+	tests := []struct {
+		name   string
+		dir    string
+		prefix string
+		expect error
```
Comment (neolit123) [inline pkg/fs/fs_test.go]: in some cases it is better to have a `expectedError bool` instead of matching the error.

File: pkg/fs/fs_test.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+			expect: nil,
+		},
+		{
+			name:   "dir is nil",
+			dir:    "",
+			prefix: "tmp",
+			expect: nil,
+		},
+		{
+			name:   "dir isn't exist",
```
Comment (neolit123) [inline pkg/fs/fs_test.go]: isn't -> doesn't

File: pkg/fs/fs_test.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+		expect error
+	}{
+		{
+			name:   "all is right",
+			dir:    ".",
+			prefix: "tmp",
+			expect: nil,
+		},
+		{
+			name:   "dir is nil",
```
Comment (neolit123) [inline pkg/fs/fs_test.go]: `dir` is not really `nil` it is an empty string.

File: pkg/fs/fs_test.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+			expect: nil,
+		},
+		{
+			name:   "dir isn't exist",
+			dir:    "/abc",
+			prefix: "tmp",
+			expect: errors.New("isn't nil"),
+		},
+		{
+			name:   "prefix is nil",
```
Comment (neolit123) [inline pkg/fs/fs_test.go]: `prefix` is not really `nil` it is an empty string.

File: pkg/fs/fs_test.go [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+			dir:    ".",
+			prefix: "",
+			expect: nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			name, err := TempDir(tt.dir, tt.prefix)
+			os.Remove(name)
+			if (nil == err) != (nil == tt.expect) {
```
Comment (neolit123) [inline pkg/fs/fs_test.go]:
```
if (err != nil) != tt.expectError {
...
```

File: pkg/fs/fs_test.go [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+		hostPath string
+		expect   bool
+	}{
+		{
+			name:     "is absolute path",
+			hostPath: "/home/test",
+			expect:   true,
+		},
+		{
+			name:     "isn't absolute path",
```
Comment (neolit123) [inline pkg/fs/fs_test.go]: `is relative path`

File: pkg/fs/fs_test.go [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+			hostPath: "./test",
+			expect:   false,
+		},
+		{
+			name:     "isn't absolute path(~)",
```
Comment (neolit123) [inline pkg/fs/fs_test.go]: `is a home path`

File: pkg/fs/fs_test.go [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+			}
+		})
+	}
+}
+
+func TestIsAbs(t *testing.T) {
+	tests := []struct {
+		name     string
+		hostPath string
+		expect   bool
```
Comment (neolit123) [inline pkg/fs/fs_test.go]: expected

File: pkg/fs/fs_test.go [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+		t.Run(tt.name, func(t *testing.T) {
+			ret := IsAbs(tt.hostPath)
+			if ret != tt.expect {
+				t.Errorf("expect: %v, got: %v", tt.expect, ret)
+			}
+		})
+	}
+}
+
+func TestCopy(t *testing.T) {
```
Comment (neolit123) [inline pkg/fs/fs_test.go]: see the comments for TestTempDir above.

File: pkg/fs/fs_test.go [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+	defer os.Remove("./test.txt")
+	defer os.RemoveAll("./test")
```
Comment (neolit123) [inline pkg/fs/fs_test.go]: tests should not use relative paths, but instead write to https://golang.org/pkg/io/ioutil/#TempDir

File: pkg/fs/fs_test.go [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+			err := Copy(tt.src, tt.dst)
+			if (nil == err) != (nil == tt.expect) {
+				t.Errorf("expect: %v, got: %v", tt.expect, err)
+			}
+		})
+	}
+}
+
+func TestCopyFile(t *testing.T) {
+	defer os.Remove("./test.txt")
```
Comment (neolit123) [inline pkg/fs/fs_test.go]: [1]

File: pkg/fs/fs_test.go [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+	tests := []struct {
+		name   string
+		src    string
+		dst    string
+		expect error
```
Comment (neolit123) [inline pkg/fs/fs_test.go]: similar comments as TestCopy and TestTempDir

File: pkg/fs/fs_test.go [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+			prefix: "",
+			expect: nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			name, err := TempDir(tt.dir, tt.prefix)
+			os.Remove(name)
+			if (nil == err) != (nil == tt.expect) {
+				t.Errorf("expect: %v, got: %v", tt.expect, err)
```
Comment (neolit123) [inline pkg/fs/fs_test.go]: `expected:...`

File: pkg/fs/fs_test.go [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+			name:     "isn't absolute path(~)",
+			hostPath: "~/test",
+			expect:   false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ret := IsAbs(tt.hostPath)
+			if ret != tt.expect {
+				t.Errorf("expect: %v, got: %v", tt.expect, ret)
```
Comment (neolit123) [inline pkg/fs/fs_test.go]: `expected:...`

File: pkg/fs/fs_test.go [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+			src:    "./fs.go",
+			dst:    "./test.txt",
+			expect: nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := Copy(tt.src, tt.dst)
+			if (nil == err) != (nil == tt.expect) {
+				t.Errorf("expect: %v, got: %v", tt.expect, err)
```
Comment (neolit123) [inline pkg/fs/fs_test.go]: `expected:...`

File: pkg/fs/fs_test.go [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+			src:    "./fs.go",
+			dst:    "./test.txt",
+			expect: nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := CopyFile(tt.src, tt.dst)
+			if (nil == err) != (nil == tt.expect) {
+				t.Errorf("expect: %v, got: %v", tt.expect, err)
```
Comment (neolit123) [inline pkg/fs/fs_test.go]: `expected:...`

File: pkg/fs/fs_test.go [hunk 17]
```diff
@@ -0,0 +1,229 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (tao12345666333) [inline pkg/fs/fs_test.go]: Can be updated to 2020 by the way :smile:

File: pkg/fs/fs_test.go [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+		},
+		{
+			name:   "src is dir",
+			src:    ".",
+			dst:    "./test.txt",
+			expect: errors.New("isn't nil"),
+		},
+		{
+			name:   "src is nil",
+			src:    "",
```
Comment (tao12345666333) [inline pkg/fs/fs_test.go]: same as neolit123's comments above.

File: pkg/fs/fs_test.go [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+
+package fs
+
+import (
+	"io/ioutil"
+	"os"
+	"testing"
+)
+
+func TestTempDir(t *testing.T) {
```
Comment (BenTheElder) [inline pkg/fs/fs_test.go]: this doesn't seem very useful, this unit test is just testing `os.TempDir` ...

if we want to unit test this, we need to retool so we can actually test the behavior we're alterting

Comment (BenTheElder) [reply pkg/fs/fs_test.go]: we should not unit test the go standard library, they have their own unit tests

File: pkg/fs/fs_test.go [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+			name, err := TempDir(tt.dir, tt.prefix)
+			os.Remove(name)
+			if (nil != err) != tt.expectedError {
+				t.Errorf("expectedError: %v, got: %v", tt.expectedError, err)
+			}
+		})
+	}
+}
+
+func TestIsAbs(t *testing.T) {
```
Comment (BenTheElder) [inline pkg/fs/fs_test.go]: again we only seem to be testing the standard library behavior instead of our special behavior

**Comment by BenTheElder** [comment]
I do think this package should have tests, but only if they're actually testing the altered behavior we have. As far as I can tell this PR currently just tests the standard library behavior which is redundant to go itself.

**Pull request closed without merging**